qiime2/sdk/usage.py

```diff
@@ -0,0 +1,368 @@
+        return factory
+
+    def _add_outputs_to_scope(self, outputs, derived_outputs):
+        outputs.validate_derived(derived_outputs)
+        for ref, result in derived_outputs.items():
+            self._push_record(ref, value=result)
+
+    def _push_record(self, ref, factory=None, value=None):
+        return self._scope.push_record(
+            ref=ref, factory=factory, value=value,
+            assert_has_line_matching=self._assert_has_line_matching_)
+
+    def _get_record(self, id):
+        return self._scope.get_record(id)
+
+    def _get_records(self):
+        return self._scope.records
+
+
+class DiagnosticUsage(Usage):
+    def __init__(self):
+        super().__init__()
+        self._recorder = []
+
+    def _action_(self, action, input_opts, output_opts):
+        self._recorder.append({
+            'type': 'action',
+            'action': action,
+            'input_opts': input_opts,
+            'output_opts': output_opts,
+        })
+        return output_opts
+
+    def _comment_(self, text):
+        self._recorder.append({
+            'type': 'comment',
+            'text': text,
+        })
+
+    def _assert_has_line_matching_(self, scope_id, label, path, expression):
+        self._recorder.append({
+            'type': 'assert_has_line_matching',
+            'scope_id': scope_id,
+            'label': label,
+            'path': path,
+            'expression': expression,
+        })
+
+    def _factory_override(self, ref, factory):
+        return lambda: ref
+
+
+class ExecutionUsage(Usage):
+    def _comment_(self, text):
+        pass
+
+    def _action_(self, action: UsageAction,
+                 input_opts: dict, output_opts: dict):
+        action_f, _ = action.get_action()
+        results = action_f(**input_opts)
+        return {k: getattr(results, v) for k, v in output_opts.items()}
+
+    def _assert_has_line_matching_(self, scope_id, label, path, expression):
+        data = self._get_record(scope_id).result
+        with tempfile.TemporaryDirectory(prefix='q2-exc-usage-') as temp_dir:
+            fp = data.save(os.path.join(temp_dir, str(scope_id)))
+            with zipfile.ZipFile(fp, 'r') as zip_temp:
+                for fn in zip_temp.namelist():
+                    if re.match(path, fn):
+                        path = fn
+                        break  # Will only match on first hit
+
+                with zip_temp.open(path) as file_temp:
+                    target = file_temp.read().decode('utf-8')
+                    match = re.search(expression, target,
+                                      flags=re.MULTILINE)
+                    if match is None:
+                        raise ValueError(
+                            'Expression %r not found in %s.' %
+                            (expression, path))
+
+    def _merge_metadata_(self, ref, records):
+        def closure_fn():
+            # Note: resolved records should be guaranteed at this point
```

Comment:
I don't think I understand why this is true, so I must be missing some part of the lifecycle. Wouldn't these records still unresolved? Nothing else has looked at them yet.

Reply:
The closure returned here is deferred until here:

https://github.com/qiime2/qiime2/pull/523/files#diff-50eb2f5e696918079827fa682a7f9fd0R257-R258

The records are resolved in the order that they are put into the scope, so any dependencies are resolved prior.